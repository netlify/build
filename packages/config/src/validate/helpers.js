'use strict'

const isPlainObj = require('is-plain-obj')

const isArrayOfObjects = function (value) {
  return Array.isArray(value) && value.every(isPlainObj)
}

const isArrayOfStrings = function (value) {
  return Array.isArray(value) && value.every(isString)
}

const isString = function (value) {
  return typeof value === 'string'
}

// Check an object valid properties, including legacy ones
const validProperties = function (propNames, legacyPropNames) {
  return {
    check: (value) => checkValidProperty(value, [...propNames, ...legacyPropNames]),
    message: `has unknown properties. Valid properties are:
${propNames.map((propName) => `  - ${propName}`).join('\n')}`,
  }
}

const checkValidProperty = function (value, propNames) {
  return Object.keys(value).every((propName) => propNames.includes(propName))
}

const functionsDirectoryCheck = {
  formatInvalid: ({ functionsDirectory } = {}) => ({ functions: { directory: functionsDirectory } }),
  propertyName: 'functions.directory',
}

module.exports = {
  isArrayOfObjects,
  isArrayOfStrings,
  isString,
  validProperties,
  functionsDirectoryCheck,
}
