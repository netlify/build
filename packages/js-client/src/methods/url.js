import queryString from 'qs'

import { getRequestParams } from './params.js'

// Replace path parameters and query parameters in the URI, using the OpenAPI
// definition
export const getUrl = function ({ path, parameters }, basePath, requestParams) {
  const url = `${basePath}${path}`
  const urlA = addPathParams(url, parameters, requestParams)
  const urlB = addQueryParams(urlA, parameters, requestParams)
  return urlB
}

const addPathParams = function (url, parameters, requestParams) {
  const pathParams = getRequestParams(parameters.path, requestParams, 'path variable')
  return Object.entries(pathParams).reduce(addPathParam, url)
}

const addPathParam = function (url, [name, value]) {
  return url.replace(`{${name}}`, value)
}

const addQueryParams = function (url, parameters, requestParams) {
  const queryParams = getRequestParams(parameters.query, requestParams, 'query variable')

  if (Object.keys(queryParams).length === 0) {
    return url
  }

  return `${url}?${queryString.stringify(queryParams, { arrayFormat: 'brackets' })}`
}
