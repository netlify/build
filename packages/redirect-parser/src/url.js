// Check if a field is valid redirect URL
export const isUrl = function (pathOrUrl) {
  return SCHEMES.some((scheme) => pathOrUrl.startsWith(scheme))
}

const SCHEMES = ['http://', 'https://']
