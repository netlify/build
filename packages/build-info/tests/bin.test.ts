import { fileURLToPath } from 'url'

import { execa, execaNode } from 'execa'
import { afterEach, expect, test } from 'vitest'

import { createFixture } from './helpers.js'

const FIXTURES_ABSOLUTE_PATH = fileURLToPath(new URL('fixtures', import.meta.url))

afterEach(async ({ cleanup }) => await cleanup?.())

const runBinary = (...args: string[]) => {
  if (process.env.CI) {
    const binary = fileURLToPath(new URL('../bin.js', import.meta.url))
    return execaNode(binary, args)
  }
  const binary = fileURLToPath(new URL('../src/node/bin.ts', import.meta.url))
  return execa('node', ['--loader=ts-node/esm', '--no-warnings', binary, ...args])
}

test('CLI --help flag', async () => {
  const { stdout } = await runBinary('--help')

  // locally we run the typescript binary but the snapshot is run in CI as well
  expect(stdout.replace(/bin\.ts/gm, 'bin.js')).toMatchSnapshot()
})

test('CLI prints js-workspaces and frameworks in JSON format', async () => {
  const { stdout } = await runBinary(
    `${FIXTURES_ABSOLUTE_PATH}/js-workspaces/packages/gatsby-site`,
    '--rootDir',
    `${FIXTURES_ABSOLUTE_PATH}/js-workspaces`,
  )
  const { jsWorkspaces, frameworks } = JSON.parse(stdout)
  expect(jsWorkspaces.isRoot).toBe(false)
  expect(jsWorkspaces.packages.length).toBe(2)
  expect(frameworks.length).toBe(1)
})

test('CLI does not print js-workspaces if given a project without it', async (ctx) => {
  const fixture = await createFixture('pnpm-simple')
  ctx.cleanup = fixture.cleanup
  const { stdout } = await runBinary(fixture.cwd)
  expect(stdout).toMatchInlineSnapshot(`
    "{
      \\"packageManager\\": {
        \\"name\\": \\"pnpm\\",
        \\"installCommand\\": \\"pnpm install\\",
        \\"runCommand\\": \\"pnpm run\\",
        \\"lockFile\\": \\"pnpm-lock.yaml\\",
        \\"forceEnvironment\\": \\"NETLIFY_USE_PNPM\\"
      },
      \\"frameworks\\": [],
      \\"buildSystems\\": []
    }"
  `)
  const { jsWorkspaces } = JSON.parse(stdout)
  expect(jsWorkspaces).toBe(undefined)
})

test('CLI works if given a relative path and no rootDir', async (ctx) => {
  const fixture = await createFixture('js-workspaces')
  ctx.cleanup = fixture.cleanup
  const { stdout } = await runBinary(fixture.cwd)
  const { jsWorkspaces, frameworks } = JSON.parse(stdout)
  expect(jsWorkspaces).not.toBe(undefined)
  expect(frameworks.length).toBe(1)
})

test('CLI prints an empty array if no frameworks are found', async (ctx) => {
  const fixture = await createFixture('empty')
  ctx.cleanup = fixture.cleanup
  const { stdout } = await runBinary(fixture.cwd)
  expect(JSON.parse(stdout)).toMatchInlineSnapshot(`
    {
      "buildSystems": [],
      "frameworks": [],
    }
  `)
  const { frameworks } = JSON.parse(stdout)
  expect(frameworks).toEqual([])
})
