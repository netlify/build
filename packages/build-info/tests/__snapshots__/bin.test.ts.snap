// Vitest Snapshot v1

exports[`CLI --help flag 1`] = `
"bin.js [projectDir]

Print relevant build information from a project.

Options:
  --help          Show help                                            [boolean]
  --version       Show version number                                  [boolean]
  --rootDir       The root directory of the project if different from projectDir
                                                                        [string]
  --featureFlags  comma separated list of feature flags                 [string]"
`;
